precompiles/parachain-staking/src/lib.rs

```diff
@@ -57,469 +62,635 @@ where
-		let is_candidate = parachain_staking::Pallet::<Runtime>::is_candidate(&candidate.into());
+		let collator_candidate_count = input.read_u32()?;
 
-		log::trace!(target: "staking-precompile", "Result from pallet is {:?}", is_candidate);
+		// Build call with origin.
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+		let call = parachain_staking::Call::<Runtime>::leave_candidates(collator_candidate_count);
 
-		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
-			<Runtime as frame_system::Config>::DbWeight::get().read,
-		);
+		// Dispatch call (if enough gas).
+		let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+			Some(origin).into(),
+			call,
+			gasometer.remaining_gas()?,
+		)?;
+		gasometer.record_cost(used_gas)?;
 
+		// Build output.
 		Ok(PrecompileOutput {
 			exit_status: ExitSucceed::Returned,
-			cost: gas_consumed,
-			output: bool_to_solidity_bytes(is_candidate),
-			logs: Default::default(),
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: vec![],
 		})
 	}
 
-	fn is_selected_candidate(input: &[u8]) -> Result<PrecompileOutput, ExitError> {
-		// parse the address
-		let candidate = H160::from_slice(&input[12..32]);
+	fn go_offline(
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> Result<PrecompileOutput, ExitError> {
+		let mut gasometer = Gasometer::new(target_gas);
 
-		log::trace!(
-			target: "staking-precompile",
-			"Checking whether {:?} is a selected collator",
-			candidate
-		);
+		// Build call with origin.
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+		let call = parachain_staking::Call::<Runtime>::go_offline();
 
-		// fetch data from pallet
-		let is_selected =
-			parachain_staking::Pallet::<Runtime>::is_selected_candidate(&candidate.into());
+		// Dispatch call (if enough gas).
+		let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+			Some(origin).into(),
+			call,
+			gasometer.remaining_gas()?,
+		)?;
+		gasometer.record_cost(used_gas)?;
 
-		log::trace!(target: "staking-precompile", "Result from pallet is {:?}", is_selected);
+		// Build output.
+		Ok(PrecompileOutput {
+			exit_status: ExitSucceed::Returned,
+			cost: gasometer.used_gas(),
+			output: vec![],
+			logs: vec![],
+		})
+	}
 
-		let gas_consumed = <Runtime as pallet_evm::Config>::GasWeightMapping::weight_to_gas(
-			<Runtime as frame_system::Config>::DbWeight::get().read,
-		);
+	fn go_online(
+		target_gas: Option<u64>,
+		context: &Context,
+	) -> Result<PrecompileOutput, ExitError> {
+		let mut gasometer = Gasometer::new(target_gas);
 
+		// Build call with origin.
+		let origin = Runtime::AddressMapping::into_account_id(context.caller);
+		let call = parachain_staking::Call::<Runtime>::go_online();
+
+		// Dispatch call (if enough gas).
+		let used_gas = RuntimeHelper::<Runtime>::try_dispatch(
+			Some(origin).into(),
+			call,
+			gasometer.remaining_gas()?,
+		)?;
+		gasometer.record_cost(used_gas)?;
```

Comment:
This is pretty repetitive. Previously the dispatch code only appeared once. I definitely like how much cleaner the `match` on the function selector looks now, but is there a way to not duplicate the code so much?

Reply:
Yes, this could be abstracted into its own function and put in `precompile-utils`. I don't think I'll have time to do this anytime soon, it could be in a follow up?